src/client/game.ts

```diff
@@ -104,79 +108,74 @@ export default class ClientGame extends GameLoop {
   }
 
   serverMsg(data: any): void {
-    if (!this.running || this.my_id === undefined) return;
+    if (!this.running) return;
 
     const message = deserializeSTC(data);
 
     if (this.my_id in message.inputAck) {
       this.inputHistory.discard_front_until(message.inputAck[this.my_id]);
     }
 
-    this.update_players(message.state);
-    this.update_enemies(message.state);
+    const prevState = this.states.last_elem();
+    const newState = message.state;
 
-    this.states.reset(message.state, message.stateNum);
+    this.update_player_sprites(prevState, newState);
+    this.update_enemy_sprites(prevState, newState);
+
+    this.states.reset(newState, message.stateNum);
   }
 
-  update_players(state: State): void {
+  update_player_sprites(prevState: State | undefined, newState: State): void {
     // spawn new players
-    for (const player of Object.values(state.players)) {
+    for (const player of Object.values(newState.players)) {
       if (this.player_list[player.id] === undefined) {
-        // the positions of the players do not matter since they will be corrected by the server
-        // TODO: we should probably think of a better solution in the future
-        this.add_character(
-          PLAYER_SPAWN_X_MIN,
-          PLAYER_SPAWN_Y_MIN,
-          PLAYER_SCALE,
-          PLAYER_SPRITE,
-          player.id,
-        );
-
-        if (player.id === this.my_id && this.my_id !== undefined) {
+        this.add_character(0, 0, PLAYER_SIZE, PLAYER_SPRITE, player.id);
+
+        if (player.id === this.my_id) {
           this.my_sprite = this.player_list[this.my_id];
         }
-      } else {
-        this.decide_direction(player.id);
-        const p = this.states.last_elem()!.players[player.id];
-        this.player_list[player.id].x = p.position.x;
-        this.player_list[player.id].y = p.position.y;
       }
+
+      this.decide_direction(prevState, newState, player.id);
+      this.player_list[player.id].x = LOGICAL_TO_PIXELS(player.position.x);
+      this.player_list[player.id].y = LOGICAL_TO_PIXELS(player.position.y);
     }
   }
 
-  update_enemies(state: State): void {
-    this.remove_enemies();
+  update_enemy_sprites(prevState: State | undefined, newState: State): void {
+    this.remove_enemy_sprites(newState);
 
-    for (const enemy of Object.values(state.enemies)) {
+    for (const enemy of Object.values(newState.enemies)) {
       if (this.enemy_list[enemy.id] === undefined) {
-        this.add_enemy(enemy.x, enemy.y, ENEMY_SCALE, ENEMY_SPRITE, enemy.id);
-      } else {
-        this.enemy_list[enemy.id].x = this.states.last_elem()!.enemies[
-          enemy.id
-        ].position.x;
-        this.enemy_list[enemy.id].y = this.states.last_elem()!.enemies[
-          enemy.id
-        ].position.y;
+        this.add_enemy(0, 0, ENEMY_SIZE, ENEMY_SPRITE, enemy.id);
```

Comment:
borde sätta positionen korrekt, alternativ göra om add_enemy så att den inte tar in en position